engine/consensus/compliance/engine.go

```diff
@@ -149,9 +173,108 @@ func (e *Engine) ProcessLocal(event interface{}) error {
 // Process processes the given event from the node with the given origin ID in
 // a blocking manner. It returns the potential processing error when done.
 func (e *Engine) Process(originID flow.Identifier, event interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(originID, event)
-	})
+	e.pendingEventSink <- &Event{
+		OriginID: originID,
+		Msg:      event,
+	}
+	return nil
+}
+
+// processEvents is processor of pending events which drives events from networking layer to business logic in `Core`.
+// Effectively consumes messages from networking layer and dispatches them into corresponding sinks which are connected with `Core`.
+// Should be run as a separate goroutine.
+func (e *Engine) processEvents() {
+	// takes pending event from one of the queues
+	// nil sink means nothing to send, this prevents blocking on select
+	fetchEvent := func() (*Event, EventSink, *fifoqueue.FifoQueue) {
+		if val, ok := e.pendingBlocks.Front(); ok {
+			return val.(*Event), e.blockSink, e.pendingBlocks
+		}
+		if val, ok := e.pendingVotes.Front(); ok {
+			return val.(*Event), e.voteSink, e.pendingVotes
+		}
```

Comment:
I am not sure, maybe it needs to be revisited

Reply:
From my perspective, prioritize blocks for the non-byzantine case is great 
* blocks are relatively rare
* if a node is behind, it should catch up first 
   * thereby, it avoids processing many votes for a view that has already passed (but the node might not be aware, if it doesn't catch up first)

regarding the byzantine case:
* processing votes and blocks can actually happen concurrently. The [HotStuff state machine](https://github.com/onflow/flow-go/blob/bf193a71552154209094382382865c5621996704/docs/StateMachine.png) does not really care for votes, it only cares for QCs. So instead of funnelling all Votes through the State machine, because some vote might reach the critical threshold for a QC to be produced, we can just notify the hotstuff state machine on the QC.
* taking it one step further, we can actually process multiple votes concurrently (as they are not state changing by themselves). 

This would reduce the attack surface for byzantine nodes to spam honest nodes with blocks to suppress vote processing. Furthermore, it would significantly speed up consensus, as sequential vote processing is time consuming.

Reply:
Hence, my recommendation:
* lets stick with the current implementation and revisit when we make vote processing concurrent